tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs

```diff
@@ -89,87 +89,116 @@ public static void Initialize()
+                    // while the app is shutting down, which results in flaky tests due to the short-
+                    // lived nature of our apps. This appears to be a bug in the runtime (although
+                    // we haven't yet confirmed that). Calling the `ToUuid()` method uses an MD5
+                    // hash which calls into the native library, triggering the load.
+                    _ = string.Empty.ToUUID();
                 }
                 catch (Exception ex)
                 {
-                    Log.Error(ex, "Error sending CallTarget integration definitions to native library");
+                    Log.Error(ex, "Error triggering eager OpenSSL load");
                 }
+#endif
+                LifetimeManager.Instance.AddShutdownTask(RunShutdown);
 
-                TelemetryFactory.Metrics.RecordDistributionSharedInitTime(MetricTags.InitializationComponent.CallTargetDefsPinvoke, sw.ElapsedMilliseconds);
-                sw.Restart();
+                Log.Debug("Initialization finished.");
 
-                InitializeTracer(sw);
+                TelemetryFactory.Metrics.RecordDistributionSharedInitTime(MetricTags.InitializationComponent.Total, swTotal.ElapsedMilliseconds);
             }
-
-#if NETSTANDARD2_0 || NETCOREAPP3_1
-            try
+            catch (Exception ex) when (IsInSsi())
             {
-                // On .NET Core 2.0-3.0 we see an occasional hang caused by OpenSSL being loaded
-                // while the app is shutting down, which results in flaky tests due to the short-
-                // lived nature of our apps. This appears to be a bug in the runtime (although
-                // we haven't yet confirmed that). Calling the `ToUuid()` method uses an MD5
-                // hash which calls into the native library, triggering the load.
-                _ = string.Empty.ToUUID();
+                // if we're in SSI we want to be as safe as possible, so catching to avoid the possibility of a crash
+                try
+                {
+                    Log.Error(ex, "Error in Datadog.Trace.ClrProfiler.Managed.Loader.Startup.Startup(). Functionality may be impacted.");
+                }
+                catch
+                {
+                    // Swallowing any errors here, as something went _very_ wrong, even with logging
+                }
             }
-            catch (Exception ex)
+
+            static bool IsInSsi()
             {
-                Log.Error(ex, "Error triggering eager OpenSSL load");
+                try
+                {
+                    // Not using the ReadEnvironmentVariable method here to avoid logging (which could cause a crash itself)
+                    return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("DD_INJECTION_ENABLED"));
+                }
+                catch
+                {
+                    // sigh, nothing works, _pretend_ we're in SSI so that we don't crash the app
```

Comment:
"pretend we're in SSI" but return `false`? seems contradictory.

Reply:
Good catch, it should return `true`, thanks 🙈